

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>5.2.1. Recursive DBScan &mdash; Autometa 2.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="5.3. Common" href="../Common/common.html" />
    <link rel="prev" title="5.2. Binning" href="binning.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Autometa
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../install.html">1. Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../databases.html">2. Databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../running-autometa.html">3. Running Autometa</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examining-results.html">4. Examining Results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../scripts.html">5. Scripts</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../main.html">5.1. Main</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="binning.html">5.2. Binning</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">5.2.1. Recursive DBScan</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../Common/common.html">5.3. Common</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Config/config.html">5.4. Config</a></li>
<li class="toctree-l2"><a class="reference internal" href="../datasets.html">5.5. Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Taxonomy/taxonomy.html">5.6. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Validation/validation.html">5.7. Validation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contribution-guidelines.html">6. Contribution Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">7. REST API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">8. License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Autometa</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../scripts.html">5. Scripts</a> &raquo;</li>
        
          <li><a href="binning.html">5.2. Binning</a> &raquo;</li>
        
      <li>5.2.1. Recursive DBScan</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/Scripts/Binning/recursive_dbscan.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="recursive-dbscan">
<h1>5.2.1. Recursive DBScan<a class="headerlink" href="#recursive-dbscan" title="Permalink to this headline">¶</a></h1>
<p>Scanning, scanning, scanning …………</p>
<span class="target" id="module-recursive_dbscan"></span><dl class="function">
<dt id="recursive_dbscan.add_metrics">
<code class="sig-prename descclassname">recursive_dbscan.</code><code class="sig-name descname">add_metrics</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">markers_df</em>, <em class="sig-param">domain='bacteria'</em><span class="sig-paren">)</span><a class="headerlink" href="#recursive_dbscan.add_metrics" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds the completeness and purity metrics to each respective contig in df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Description of parameter <cite>df</cite>.</p></li>
<li><p><strong>markers_df</strong> (<em>pd.DataFrame</em>) – wide format, i.e. index=contig cols=[marker,marker,…]</p></li>
<li><p><strong>domain</strong> (<em>str</em>) – Kingdom to determine metrics (the default is ‘bacteria’).
choices=[‘bacteria’,’archaea’]</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>DataFrame with added columns - ‘completeness’ and ‘purity’</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="recursive_dbscan.binning">
<code class="sig-prename descclassname">recursive_dbscan.</code><code class="sig-name descname">binning</code><span class="sig-paren">(</span><em class="sig-param">master</em>, <em class="sig-param">markers</em>, <em class="sig-param">domain='bacteria'</em>, <em class="sig-param">completeness=20.0</em>, <em class="sig-param">purity=90.0</em>, <em class="sig-param">taxonomy=True</em>, <em class="sig-param">method='DBSCAN'</em>, <em class="sig-param">reverse=True</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#recursive_dbscan.binning" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform clustering of contigs by provided <cite>method</cite> and use metrics to
filter clusters that should be retained via <cite>completeness</cite> and <cite>purity</cite>
thresholds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>master</strong> (<em>pd.DataFrame</em>) – <p>index=contig,
cols=</p>
<blockquote>
<div><p>embedded-kmers are cols ‘x’,’y’ and ‘z’
taxa cols should be present if <cite>taxonomy</cite> is True.
i.e. [taxid,superkingdom,phylum,class,order,family,genus,species]</p>
</div></blockquote>
</p></li>
<li><p><strong>markers</strong> (<em>pd.DataFrame</em>) – wide format, i.e. index=contig cols=[marker,marker,…]</p></li>
<li><p><strong>domain</strong> (<em>str</em>) – Kingdom to determine metrics (the default is ‘bacteria’).
choices=[‘bacteria’,’archaea’]</p></li>
<li><p><strong>completeness</strong> (<em>float</em>) – Description of parameter <cite>completeness</cite> (the default is 20.).</p></li>
<li><p><strong>purity</strong> (<em>float</em>) – Description of parameter <cite>purity</cite> (the default is 90.).</p></li>
<li><p><strong>taxonomy</strong> (<em>bool</em>) – Split canonical ranks and subset based on rank then attempt to find clusters (the default is True).
taxonomic_levels = [superkingdom,phylum,class,order,family,genus,species]</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Clustering <cite>method</cite> (the default is ‘DBSCAN’).
choices = [‘DBSCAN’,’HDBSCAN’]</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – True - [superkingdom,phylum,class,order,family,genus,species]
False - [species,genus,family,order,class,phylum,superkingdom]</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – log stats for each recursive_dbscan clustering iteration</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>master with [‘cluster’,’completeness’,’purity’] columns added</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><a class="reference internal" href="../Common/exceptions.html#exceptions.RecursiveDBSCANError" title="exceptions.RecursiveDBSCANError"><strong>RecursiveDBSCANError</strong></a> – No marker information is availble for contigs to be binned.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="recursive_dbscan.get_clusters">
<code class="sig-prename descclassname">recursive_dbscan.</code><code class="sig-name descname">get_clusters</code><span class="sig-paren">(</span><em class="sig-param">master_df</em>, <em class="sig-param">markers_df</em>, <em class="sig-param">domain='bacteria'</em>, <em class="sig-param">completeness=20.0</em>, <em class="sig-param">purity=90.0</em>, <em class="sig-param">method='DBSCAN'</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#recursive_dbscan.get_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Find best clusters retained after applying <cite>completeness</cite> and <cite>purity</cite> filters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>master_df</strong> (<em>pd.DataFrame</em>) – <p>index=contig,
cols=</p>
<blockquote>
<div><p>embedded-kmers are cols ‘x’,’y’ and ‘z’</p>
</div></blockquote>
</p></li>
<li><p><strong>markers_df</strong> (<em>pd.DataFrame</em>) – wide format, i.e. index=contig cols=[marker,marker,…]</p></li>
<li><p><strong>domain</strong> (<em>str</em>) – Kingdom to determine metrics (the default is ‘bacteria’).
choices=[‘bacteria’,’archaea’].</p></li>
<li><p><strong>completeness</strong> (<em>float</em>) – <cite>completeness</cite> threshold to retain cluster (the default is 20.).</p></li>
<li><p><strong>purity</strong> (<em>float</em>) – <cite>purity</cite> threshold to retain cluster (the default is 90.).</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Description of parameter <cite>method</cite> (the default is ‘DBSCAN’).
choices = [‘DBSCAN’,’HDBSCAN’]</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – log stats for each recursive_dbscan clustering iteration</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>master_df</cite> with [‘cluster’,’completeness’,’purity’] columns added</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="recursive_dbscan.recursive_dbscan">
<code class="sig-prename descclassname">recursive_dbscan.</code><code class="sig-name descname">recursive_dbscan</code><span class="sig-paren">(</span><em class="sig-param">table</em>, <em class="sig-param">markers_df</em>, <em class="sig-param">domain</em>, <em class="sig-param">completeness_cutoff</em>, <em class="sig-param">purity_cutoff</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">method='DBSCAN'</em><span class="sig-paren">)</span><a class="headerlink" href="#recursive_dbscan.recursive_dbscan" title="Permalink to this definition">¶</a></dt>
<dd><p>Carry out DBSCAN, starting at eps=0.3 and continuing until there is just one
group.</p>
<p>Break conditions to speed up pipeline:
Give up if we’ve got up to eps 1.3 and still no complete and pure clusters
Often when you start at 0.3 there are zero complete and pure clusters, because
the groups are too small. Later, some are found as the groups enlarge enough, but
after it becomes zero again, it is a lost cause and we may as well stop. On the
other hand, sometimes we never find any groups, so perhaps we should give up if
by EPS 1.3 we never find any complete/pure groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table</strong> (<em>pd.DataFrame</em>) – Contigs with embedded k-mer frequencies as [‘x’,’y’,’z’] columns and
optionally ‘coverage’ column</p></li>
<li><p><strong>markers_df</strong> (<em>pd.DataFrame</em>) – wide format, i.e. index=contig cols=[marker,marker,…]</p></li>
<li><p><strong>domain</strong> (<em>str</em>) – Kingdom to determine metrics (the default is ‘bacteria’).
choices=[‘bacteria’,’archaea’]</p></li>
<li><p><strong>completeness_cutoff</strong> (<em>float</em>) – <cite>completeness_cutoff</cite> threshold to retain cluster (the default is 20.0).</p></li>
<li><p><strong>purity_cutoff</strong> (<em>float</em>) – <cite>purity_cutoff</cite> threshold to retain cluster (the default is 90.0).</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – log stats for each recursive_dbscan clustering iteration.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – clustering <cite>method</cite> to perform. (the default is ‘DBSCAN’).
Choices=[‘DBSCAN’,’HDBSCAN’] NOTE: ‘HDBSCAN’ is still under development.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>(pd.DataFrame(&lt;passed cutoffs&gt;), pd.DataFrame(&lt;failed cutoffs&gt;))
DataFrames consisting of contigs that passed/failed clustering
cutoffs, respectively.</p>
<dl class="simple">
<dt>DataFrame:</dt><dd><p>index = contig
columns = [‘x,’y’,’z’,’coverage’,’cluster’,’purity’,’completeness’]</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>2-tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="recursive_dbscan.run_dbscan">
<code class="sig-prename descclassname">recursive_dbscan.</code><code class="sig-name descname">run_dbscan</code><span class="sig-paren">(</span><em class="sig-param">df, eps, dropcols=['cluster', 'purity', 'completeness'], usecols=['x', 'y'], method='DBSCAN'</em><span class="sig-paren">)</span><a class="headerlink" href="#recursive_dbscan.run_dbscan" title="Permalink to this definition">¶</a></dt>
<dd><p>Run clustering via <cite>method</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Description of parameter <cite>df</cite>.</p></li>
<li><p><strong>eps</strong> (<em>float</em>) – Description of parameter <cite>eps</cite>.</p></li>
<li><p><strong>dropcols</strong> (<em>list</em>) – Drop columns in list from <cite>df</cite> (the default is [‘cluster’,’purity’,’completeness’]).</p></li>
<li><p><strong>usecols</strong> (<em>list</em>) – Use columns in list for <cite>df</cite>.
The default is [‘x’,’y’,’coverage’] if ‘coverage’ exists in df.columns.
else [‘x’,’y’,’z’].</p></li>
<li><p><strong>method</strong> (<em>str</em>) – clustering <cite>method</cite> to perform. (the default is ‘DBSCAN’).
Choices=[‘DBSCAN’,’HDBSCAN’] NOTE: ‘HDBSCAN’ is still under development</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>df</cite> with ‘cluster’ column added</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – sets <cite>usecols</cite> and <cite>dropcols</cite> may not share elements</p></li>
<li><p><strong>ValueError</strong> – Method is not an available choice: choices=[‘DBSCAN’,’HDBSCAN’]</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../Common/common.html" class="btn btn-neutral float-right" title="5.3. Common" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="binning.html" class="btn btn-neutral float-left" title="5.2. Binning" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015 - 2020, Evan R. Rees, Shaurya Chanana, Siddharth Uppal, Kyle Wolf, Jason C. Kwan
      <span class="lastupdated">
        Last updated on Mar 26, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>