

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>5.3.7. metagenome.py &mdash; Autometa 2.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="5.3.8. utilities.py" href="utilities.html" />
    <link rel="prev" title="5.3.6. markers.py" href="markers.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> Autometa
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../install.html">1. Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../databases.html">2. Databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../running-autometa.html">3. Running Autometa</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../examining-results.html">4. Examining Results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../scripts.html">5. Scripts</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../main.html">5.1. Main</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Binning/binning.html">5.2. Binning</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="common.html">5.3. Common</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="External/external.html">5.3.1. External</a></li>
<li class="toctree-l3"><a class="reference internal" href="coverage.html">5.3.2. coverage.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="exceptions.html">5.3.3. exceptions.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="kmers.html">5.3.4. kmers.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="mag.html">5.3.5. mag.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="markers.html">5.3.6. markers.py</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">5.3.7. metagenome.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="utilities.html">5.3.8. utilities.py</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../Config/config.html">5.4. Config</a></li>
<li class="toctree-l2"><a class="reference internal" href="../datasets.html">5.5. Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Taxonomy/taxonomy.html">5.6. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../Validation/validation.html">5.7. Validation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contribution-guidelines.html">6. Contribution Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">7. REST API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">8. License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Autometa</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../scripts.html">5. Scripts</a> &raquo;</li>
        
          <li><a href="common.html">5.3. Common</a> &raquo;</li>
        
      <li>5.3.7. metagenome.py</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/Scripts/Common/metagenome.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="metagenome-py">
<h1>5.3.7. metagenome.py<a class="headerlink" href="#metagenome-py" title="Permalink to this headline">¶</a></h1>
<p class="rubric">Classes</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#metagenome.Metagenome" title="metagenome.Metagenome"><code class="xref py py-obj docutils literal notranslate"><span class="pre">metagenome.Metagenome</span></code></a>(assembly, outdir, …)</p></td>
<td><p>Autometa Metagenome Class.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-metagenome"></span><p>Script containing Metagenome class for general handling of metagenome assembly</p>
<dl class="class">
<dt id="metagenome.Metagenome">
<em class="property">class </em><code class="sig-prename descclassname">metagenome.</code><code class="sig-name descname">Metagenome</code><span class="sig-paren">(</span><em class="sig-param">assembly</em>, <em class="sig-param">outdir</em>, <em class="sig-param">nucl_orfs_fpath</em>, <em class="sig-param">prot_orfs_fpath</em>, <em class="sig-param">taxonomy_fpath</em>, <em class="sig-param">fwd_reads=None</em>, <em class="sig-param">rev_reads=None</em>, <em class="sig-param">taxon_method='majority_vote'</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome" title="Permalink to this definition">¶</a></dt>
<dd><p>Autometa Metagenome Class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>assembly</strong> (<em>str</em>) – &lt;/path/to/assembled/metagenome.fasta&gt;</p></li>
<li><p><strong>outdir</strong> (<em>type</em>) – &lt;/path/to/output/directory&gt; (the default is None)</p></li>
<li><p><strong>taxon_method</strong> (<em>str</em>) – method to assign taxonomy (the default is ‘majority_vote’).
choices=[‘majority_vote’]</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="metagenome.Metagenome.nucl_orfs_fpath">
<code class="sig-name descname">nucl_orfs_fpath</code><a class="headerlink" href="#metagenome.Metagenome.nucl_orfs_fpath" title="Permalink to this definition">¶</a></dt>
<dd><p>Description of attribute <cite>nucl_orfs_fpath</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.prot_orfs_fpath">
<code class="sig-name descname">prot_orfs_fpath</code><a class="headerlink" href="#metagenome.Metagenome.prot_orfs_fpath" title="Permalink to this definition">¶</a></dt>
<dd><p>Description of attribute <cite>prot_orfs_fpath</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.taxonomy_fname">
<code class="sig-name descname">taxonomy_fname</code><a class="headerlink" href="#metagenome.Metagenome.taxonomy_fname" title="Permalink to this definition">¶</a></dt>
<dd><p>Description of attribute <cite>taxonomy_fname</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.taxonomy_fpath">
<code class="sig-name descname">taxonomy_fpath</code><a class="headerlink" href="#metagenome.Metagenome.taxonomy_fpath" title="Permalink to this definition">¶</a></dt>
<dd><p>Description of attribute <cite>taxonomy_fpath</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.taxonomy">
<code class="sig-name descname">taxonomy</code><a class="headerlink" href="#metagenome.Metagenome.taxonomy" title="Permalink to this definition">¶</a></dt>
<dd><p>index=contig cols=[taxid] may also contain lineage of taxid</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.taxonomy_assigned">
<code class="sig-name descname">taxonomy_assigned</code><a class="headerlink" href="#metagenome.Metagenome.taxonomy_assigned" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>taxonomy_fpath</cite> exists</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.orfs_called">
<code class="sig-name descname">orfs_called</code><a class="headerlink" href="#metagenome.Metagenome.orfs_called" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>nucl_orfs_fpath</cite> and <cite>prot_orfs_fpath</cite> exists</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.sequences">
<code class="sig-name descname">sequences</code><a class="headerlink" href="#metagenome.Metagenome.sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>[SeqRecord,…]</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.nseqs">
<code class="sig-name descname">nseqs</code><a class="headerlink" href="#metagenome.Metagenome.nseqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of sequences in assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.mean_gc">
<code class="sig-name descname">mean_gc</code><a class="headerlink" href="#metagenome.Metagenome.mean_gc" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean GC% of assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.size">
<code class="sig-name descname">size</code><a class="headerlink" href="#metagenome.Metagenome.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Total assembly size.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.largest_seq">
<code class="sig-name descname">largest_seq</code><a class="headerlink" href="#metagenome.Metagenome.largest_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>id of longest sequence in assembly</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.nucls">
<code class="sig-name descname">nucls</code><a class="headerlink" href="#metagenome.Metagenome.nucls" title="Permalink to this definition">¶</a></dt>
<dd><p>SeqRecords all correspond to retrieved nucleotide ORFs. [SeqRecord, …].</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="metagenome.Metagenome.prots">
<code class="sig-name descname">prots</code><a class="headerlink" href="#metagenome.Metagenome.prots" title="Permalink to this definition">¶</a></dt>
<dd><p>SeqRecords all correspond to retrieved amino-acid ORFs. [SeqRecord, …].</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.fragmentation_metric()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.describe()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.length_filter()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.call_orfs()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.orfs()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.get_kmers()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.assign_taxonomy()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.get_kingdoms()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">- self.write_ranks()</code></dt>
<dd></dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.assign_taxonomy">
<code class="sig-name descname">assign_taxonomy</code><span class="sig-paren">(</span><em class="sig-param">method</em>, <em class="sig-param">force=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.assign_taxonomy" title="Permalink to this definition">¶</a></dt>
<dd><p>Assign taxonomy to each sequence in assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>force</strong> (<em>bool</em>) – overwrite existing voting method’s file (the default is False).</p></li>
<li><p><strong>*args</strong> (<em>type</em>) – Description of parameter <cite>*args</cite>.</p></li>
<li><p><strong>**kwargs</strong> (<em>type</em>) – Description of parameter <cite>**kwargs</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ExceptionName</strong> – Why the exception is raised.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.call_orfs">
<code class="sig-name descname">call_orfs</code><span class="sig-paren">(</span><em class="sig-param">force=False</em>, <em class="sig-param">cpus=0</em>, <em class="sig-param">parallel=True</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.call_orfs" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls ORFs on Metagenome assembly.</p>
<p>(Wrapper using external executable: prodigal).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>force</strong> (<em>bool</em>) – force overwrite of existing ORFs files (the default is False).</p></li>
<li><p><strong>cpus</strong> (<em>int</em>) – Description of parameter <cite>cpus</cite> (the default is 0).</p></li>
<li><p><strong>parallel</strong> (<em>bool</em>) – Will parallelize prodigal using GNU parallel (the default is True).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(&lt;/path/to/nucls.orfs.fna&gt;, &lt;/path/to/prots.orfs.faa&gt;)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>2-tuple</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – <cite>force</cite>,`parallel` or <cite>cpus</cite> type was incorrectly supplied.</p></li>
<li><p><strong>OSError</strong> – ORF calling failed.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.fragmentation_metric">
<code class="sig-name descname">fragmentation_metric</code><span class="sig-paren">(</span><em class="sig-param">quality_measure=0.5</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.fragmentation_metric" title="Permalink to this definition">¶</a></dt>
<dd><p>Describes the quality of assembled genomes that are fragmented in
contigs of different length.</p>
<dl class="simple">
<dt>For more information see:</dt><dd><p><a class="reference external" href="http://www.metagenomics.wiki/pdf/definition/assembly/n50">http://www.metagenomics.wiki/pdf/definition/assembly/n50</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>quality_measure</strong> (<em>0 &lt; float &lt; 1</em>) – Description of parameter <cite>quality_measure</cite> (the default is .50).
I.e. default measure is N50, but could use .1 for N10 or .9 for N90</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Minimum contig length to cover <cite>quality_measure</cite> of genome (i.e. length
weighted median)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.get_kingdoms">
<code class="sig-name descname">get_kingdoms</code><span class="sig-paren">(</span><em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.get_kingdoms" title="Permalink to this definition">¶</a></dt>
<dd><p>Separate sequences by kingdom using supplied taxon assignment method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>**kwargs</strong> (<em>dict</em>) – Optional additional keyword arguments to supply to assign_taxonomy
and supply a separate NCBI_DIR</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>{‘bacteria’:MAG, …}</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>KeyError</strong> – Why the exception is raised.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.get_kmers">
<code class="sig-name descname">get_kmers</code><span class="sig-paren">(</span><em class="sig-param">kmer_size=5</em>, <em class="sig-param">multiprocess=True</em>, <em class="sig-param">out=None</em>, <em class="sig-param">normalized=None</em>, <em class="sig-param">force=False</em>, <em class="sig-param">nproc=1</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.get_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts k-mer frequencies using provided <cite>kmer_size</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kmer_size</strong> (<em>int</em>) – length of k-mer to count (the default is 5).</p></li>
<li><p><strong>normalized</strong> (<em>str</em>) – Perform Centered-log ratio normalization on counted k-mers (the default is None).
and write to provided <cite>normalized</cite> path.</p></li>
<li><p><strong>out</strong> (<em>str</em>) – Write counted k-mers to <cite>out</cite> (the default is None).</p></li>
<li><p><strong>force</strong> (<em>bool</em>) – Overwrite existing k-mers <cite>out</cite> file (the default is False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pandas.DataFrame</em> – pandas DataFrame</p></li>
<li><p><strong>TODO</strong> (<em>get_kmers should handle both files and SeqRecords…</em>)</p></li>
<li><p><strong>NOTE</strong> (<em>above TODO should be handled in kmers.py not here…</em>)</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.length_filter">
<code class="sig-name descname">length_filter</code><span class="sig-paren">(</span><em class="sig-param">out</em>, <em class="sig-param">cutoff=3000</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.length_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters sequences by length with provided cutoff.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cutoff</strong> (<em>int</em>) – Lengths above or equal to <cite>cutoff</cite> that will be retained (the default is 3000).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>autometa Metagenome object with only assembly sequences above the cutoff threshold.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#metagenome.Metagenome" title="metagenome.Metagenome">Metagenome</a></p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>TypeError</strong> – cutoff value must be a float or integer</p></li>
<li><p><strong>ValueError</strong> – cutoff value must be a positive real number</p></li>
<li><p><strong>FileExistsError</strong> – filepath consisting of sequences that passed filter already exists</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.orfs">
<code class="sig-name descname">orfs</code><span class="sig-paren">(</span><em class="sig-param">orf_type='prot'</em>, <em class="sig-param">cpus=0</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.orfs" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieves ORFs after being called from self.call_orfs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>orf_type</strong> (<em>str</em>) – format of ORFs to retrieve choices=[‘nucl’,’prot’] either nucleotide
or amino acids (the default is ‘prot’).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[SeqRecord, …]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – Invalid <cite>orf_type</cite>. Choices=[‘prot’,’nucl’]</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="metagenome.Metagenome.write_ranks">
<code class="sig-name descname">write_ranks</code><span class="sig-paren">(</span><em class="sig-param">rank='superkingdom'</em><span class="sig-paren">)</span><a class="headerlink" href="#metagenome.Metagenome.write_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Write fastas split by <cite>rank</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>rank</strong> (<em>str</em>) – <cite>rank</cite> from canonical ranks (the default is ‘superkingdom’).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>[rank_name_fp, …]</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – <cite>rank</cite> not in canonical ranks</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="utilities.html" class="btn btn-neutral float-right" title="5.3.8. utilities.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="markers.html" class="btn btn-neutral float-left" title="5.3.6. markers.py" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015 - 2020, Evan R. Rees, Shaurya Chanana, Siddharth Uppal, Kyle Wolf, Jason C. Kwan
      <span class="lastupdated">
        Last updated on Mar 26, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>