---
title: "Autometa Mock Data Report"
output: html_document
params:
  bins_path: "*.binning.main.tsv.gz"
  assembly_to_locus_path: "assembly_to_locus.txt"
  assembly_report_path: "assembly_report.txt"
  assigned_taxonomy_path: "*taxonomy.tsv.gz"
  genus: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, error=TRUE)
```

```{r message=FALSE, warning=FALSE}

packages <- c("data.table", "ggplot2", "plotly", "crosstalk", "magrittr", "DT", "ggbeeswarm", "patchwork")

for (i in packages) {
    if (!requireNamespace(i)) {
        install.packages(i)
    }
    library(i, character.only = T)
}
```

```{r eval=TRUE, include=FALSE}
params= list(
    bins_path= "*.binning.main.tsv.gz",
    assembly_to_locus_path= "assembly_to_locus.txt",
    assembly_report_path= "assembly_report.txt",
    genus= TRUE)
```



```{r read-and-clean-bins}
bins <- fread(list.files(pattern=params$bins_path,full.names = T)[[1]])

# extract locus from "contig" column which is locus_coordinates 
bins$locus = vapply(bins$contig,
                    function(x){
                        paste(strsplit(x, "_")[[1]][1:2],collapse="_")
                    },
                    FUN.VALUE = "")
```

```{r read-and-clean-assembly-to-locus}
assembly_to_locus <- fread(
    params$assembly_to_locus_path,
    sep = "/",
    header = F,
    col.names = c("assembly", "locus"))

# clean assembly id
assembly_to_locus$assembly = vapply(assembly_to_locus$assembly,
                                    function(x) {
                                        paste(strsplit(x, "_")[[1]][1:2],
                                              collapse = "_")
                                    },
                                    FUN.VALUE = "")
```


```{r read-and-clean-assembly-report, message=FALSE, warning=FALSE}
assembly_report <- fread(params$assembly_report_path)
assembly_report$actual_genus <-  gsub("\\s.*", "", assembly_report$organism_name)
assembly_report$actual_taxid <- assembly_report$taxid
assembly_report$taxid =NULL
colnames(assembly_report)[1] <- "assembly"

assembly_report=assembly_report[assembly %in% assembly_to_locus$assembly]

assembly_report = assembly_report[, 
                                  c("assembly",
                                      "refseq_category",
                                      "species_taxid",
                                      "organism_name",
                                      "assembly_level",
                                      "genome_rep",
                                      "relation_to_type_material",
                                      "actual_genus",
                                      "actual_taxid")
                                  ]
```


```{r merge-all}
bins2 <- merge(bins, assembly_to_locus, by = "locus", all.x = T, all.y = T)
bins3 <- merge(bins2, assembly_report, by = "assembly", all.x = T, all.y = T)
bins3[bins3[is.na(x_1)], x_1 := sample(1:100, length(bins3[bins3[is.na(x_1)]]$x_1) , replace=T)]
bins3[bins3[is.na(x_2)], x_2 := sample(100:100, length(bins3[bins3[is.na(x_2)]]$x_2) , replace=T)]
```


```{r}

```



## Summary

Summary information about the input genomes.

```{r summary}



colnames(assembly_report)[9] <- "taxid"

DT::datatable(z,  options = list(
    columnDefs = list(list(className = 'dt-center',  targets = "_all"))))

```



```{r plotly-crosstalk-key}
bins_copy <- data.table::copy(bins2)
bins3 <- highlight_key(bins2, ~contig)
```


```{r plotly-xy}
if (params$genus) {
    plotly_xy <- plot_ly(bins3,
                         x = ~x_1,
                         y = ~x_2,
                         color = ~genus,
                         hoverinfo ="text",
                         legendgroup='group1',
                         
                         hovertext = paste("assembly:", bins_copy$assembly,
                                           "<br> locus:", bins_copy$locus,
                                           "<br> contig:", bins_copy$contig,
                                           "<br> genus:", bins_copy$genus
                         )) %>%
        add_markers()
    
} else {
    plotly_xy <- plot_ly(bins3,
                         x = ~x_1,
                         y = ~x_2,
                         color = ~assembly,
                         hoverinfo ="text",
                         showlegend=FALSE,
                         legendgroup='group1',
                         
                         hovertext = paste("assembly:", bins_copy$assembly,
                                           "<br> locus:", bins_copy$locus,
                                           "<br> contig:", bins_copy$contig)) %>%
        add_markers()
}


```


```{r plotly-xyz}
if (params$genus) {
    plotly_xyz <- plot_ly(bins3,
                          x = ~x_1,
                          y = ~x_2,
                          z = ~coverage,
                          color = ~genus,
                          hoverinfo ="text",         showlegend=FALSE,
                          legendgroup='group1',
                          hovertext = paste("assembly:", bins_copy$assembly,
                                            "<br> locus:", bins_copy$locus,
                                            "<br> contig:", bins_copy$contig,
                                            "<br> genus:", bins_copy$genus
                          ))%>%
        add_markers()
} else {
    
    plotly_xyz <- plot_ly(bins3,
                          x = ~x_1,
                          y = ~x_2,
                          z = ~coverage,
                          color = ~assembly,
                          hoverinfo ="text",
                          showlegend=FALSE,
                          legendgroup='group1',
                          hovertext = paste("assembly:", bins_copy$assembly,
                                            "<br> locus:", bins_copy$locus,
                                            "<br> contig:", bins_copy$contig))%>%
        add_markers()
}

```

## Plots

The plots below are linked ("Brushing"/Selecting points in the 2D plot will highlight points in the 3D plot).


```{r message=FALSE, warning=FALSE, fig.height=12, fig.width=12}

fig <- subplot(plotly_xy, plotly_xyz,nrows = 2)

fig %>% layout(
    xaxis = list(domain=list(x=c(0,0.5),y=c(0,0.5))),
    
    scene = list(domain=list(x=c(0,1),y=c(0,0.5))),
    
    xaxis2 = list(domain=list(x=c(0,.5),y=c(0,0.5)))) %>%
    highlight(on = "plotly_selected", dynamic = TRUE)#, selectize = TRUE)

```










```{r plotly-crosstalk-key}
bins_copy <- data.table::copy(bins2)
bins3 <- highlight_key(bins2, ~cluster)

if (params$genus) {
    plotly_xy <- plot_ly(bins3,
                         x = ~x_1,
                         y = ~x_2,
                         color = ~cluster,
                         hoverinfo ="text",
                         legendgroup='group1',
                         
                         hovertext = paste("assembly:", bins_copy$assembly,
                                           "<br> locus:", bins_copy$locus,
                                           "<br> contig:", bins_copy$contig,
                                           "<br> genus:", bins_copy$genus
                         )) %>%
        add_markers()
    
} else {
    plotly_xy <- plot_ly(bins3,
                         x = ~x_1,
                         y = ~x_2,
                         color = ~cluster,
                         hoverinfo ="text",
                         showlegend=FALSE,
                         legendgroup='group1',
                         
                         hovertext = paste("assembly:", bins_copy$assembly,
                                           "<br> locus:", bins_copy$locus,
                                           "<br> contig:", bins_copy$contig)) %>%
        add_markers()
}




if (params$genus) {
    plotly_xyz <- plot_ly(bins3,
                          x = ~x_1,
                          y = ~x_2,
                          z = ~coverage,
                          color = ~cluster,
                          hoverinfo ="text",         showlegend=FALSE,
                          legendgroup='group1',
                          hovertext = paste("assembly:", bins_copy$assembly,
                                            "<br> locus:", bins_copy$locus,
                                            "<br> contig:", bins_copy$contig,
                                            "<br> genus:", bins_copy$genus
                          ))%>%
        add_markers()
} else {
    
    plotly_xyz <- plot_ly(bins3,
                          x = ~x_1,
                          y = ~x_2,
                          z = ~coverage,
                          color = ~cluster,
                          hoverinfo ="text",
                          showlegend=FALSE,
                          legendgroup='group1',
                          hovertext = paste("assembly:", bins_copy$assembly,
                                            "<br> locus:", bins_copy$locus,
                                            "<br> contig:", bins_copy$contig))%>%
        add_markers()
}




fig <- subplot(plotly_xy, plotly_xyz,nrows = 2)

fig %>% layout(
    xaxis = list(domain=list(x=c(0,0.5),y=c(0,0.5))),
    
    scene = list(domain=list(x=c(0,1),y=c(0,0.5))),
    
    xaxis2 = list(domain=list(x=c(0,.5),y=c(0,0.5)))) %>%
    highlight(on = "plotly_selected", dynamic = TRUE)#, selectize = TRUE)

```






























































































```{r}
if (params$genus) {
    p1 <- ggplot(bins, aes(x = genus, y = x_1)) +
        ggbeeswarm::geom_quasirandom(aes(color= genus), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    
    p2 <- ggplot(bins, aes(x = genus, y = x_2)) +
        ggbeeswarm::geom_quasirandom(aes(color= genus), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    
    
    p3 <- ggplot(bins, aes(x = genus, y = gc_content)) +
        ggbeeswarm::geom_quasirandom(aes(color= genus), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
} else {
    p1 <- ggplot(bins, aes(x = assembly, y = x_1)) +
        ggbeeswarm::geom_quasirandom(aes(color= assembly), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    
    p2 <- ggplot(bins, aes(x = assembly, y = x_2)) +
        ggbeeswarm::geom_quasirandom(aes(color= assembly), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    
    
    p3 <- ggplot(bins, aes(x = assembly, y = gc_content)) +
        ggbeeswarm::geom_quasirandom(aes(color= assembly), alpha=0.3) +
        theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
}
p1+p2+p3+ plot_layout(guides = "collect")
```
