

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>5.3.1.3. Diamond &mdash; Autometa 2.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="5.3.1.4. Hmmer" href="hmmer.html" />
    <link rel="prev" title="5.3.1.2. Bowtie" href="bowtie.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> Autometa
          

          
          </a>

          
            
            
              <div class="version">
                2.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">1. Install</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../databases.html">2. Databases</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../running-autometa.html">3. Running Autometa</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examining-results.html">4. Examining Results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../scripts.html">5. Scripts</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../main.html">5.1. Main</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Binning/binning.html">5.2. Binning</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../common.html">5.3. Common</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="external.html">5.3.1. External</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="bedtools.html">5.3.1.1. Bedtools</a></li>
<li class="toctree-l4"><a class="reference internal" href="bowtie.html">5.3.1.2. Bowtie</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">5.3.1.3. Diamond</a></li>
<li class="toctree-l4"><a class="reference internal" href="hmmer.html">5.3.1.4. Hmmer</a></li>
<li class="toctree-l4"><a class="reference internal" href="prodigal.html">5.3.1.5. Prodigal</a></li>
<li class="toctree-l4"><a class="reference internal" href="samtools.html">5.3.1.6. Samtools</a></li>
<li class="toctree-l4"><a class="reference internal" href="work_queue.html">5.3.1.7. Work queue</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../coverage.html">5.3.2. coverage.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../exceptions.html">5.3.3. exceptions.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../kmers.html">5.3.4. kmers.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../mag.html">5.3.5. mag.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../markers.html">5.3.6. markers.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../metagenome.html">5.3.7. metagenome.py</a></li>
<li class="toctree-l3"><a class="reference internal" href="../utilities.html">5.3.8. utilities.py</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../Config/config.html">5.4. Config</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../datasets.html">5.5. Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Taxonomy/taxonomy.html">5.6. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Validation/validation.html">5.7. Validation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../contribution-guidelines.html">6. Contribution Guidelines</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">7. REST API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">8. License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Autometa</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../scripts.html">5. Scripts</a> &raquo;</li>
        
          <li><a href="../common.html">5.3. Common</a> &raquo;</li>
        
          <li><a href="external.html">5.3.1. External</a> &raquo;</li>
        
      <li>5.3.1.3. Diamond</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../../_sources/Scripts/Common/External/diamond.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="diamond">
<h1>5.3.1.3. Diamond<a class="headerlink" href="#diamond" title="Permalink to this headline">¶</a></h1>
<p>File containing class and functions related to running diamond on metagenome sequences</p>
<p><strong>Running the script as stand alone module</strong>
While in the Autometa directory and with the conda environment activated, run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python -m autometa.common.external.diamond &lt;path/to/input/file.faa&gt;
&lt;/path/to/diamond/formatted/database&gt; &lt;/path/to/ncbi/prot.accession2taxid.gz&gt;
&lt;/path/to/diamond/output/file&gt; &lt;blastp or blastx&gt; --<span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
<p><strong>Inputs</strong> : Protein file (faa format)</p>
<p><strong>Returns</strong> : A dictionary of Blastp hits, with key = qseqid and value = Diamond result</p>
<p><strong>Usage and Command line options:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Usage: Retrieves blastp hits with provided input assembly <span class="o">[</span>-h<span class="o">]</span> <span class="o">[</span>--evalue EVALUE<span class="o">]</span> <span class="o">[</span>--maxtargetseqs MAXTARGETSEQS<span class="o">]</span>
                                                          <span class="o">[</span>--cpus CPUS<span class="o">]</span> <span class="o">[</span>--tmpdir TMPDIR<span class="o">]</span> <span class="o">[</span>--tmpdir TMPDIR<span class="o">]</span>
                                                          <span class="o">[</span>--top-pct TOP_PCT<span class="o">]</span> <span class="o">[</span>--force<span class="o">]</span> <span class="o">[</span>--verbose<span class="o">]</span> fasta database
                                                          acc2taxids outfile <span class="o">{</span>blastp,blastx<span class="o">}</span>

Positional arguments:
fasta                               &lt;/path/to/input/faa/file&gt;
database                            &lt;/path/to/diamond/formatted/database&gt;
acc2taxids                          &lt;/path/to/ncbi/prot.accession2taxid.gz&gt;
outfile                             &lt;/path/to/diamond/output/file&gt;
blast <span class="nb">type</span> <span class="o">{</span>blastp,blastx<span class="o">}</span>          <span class="o">[</span>blastp<span class="o">]</span>: protein -&gt; protein or <span class="o">[</span>blastx<span class="o">]</span>: Nucleic acid -&gt; protein

Optional arguments:
-h, --help                          show this <span class="nb">help</span> message and <span class="nb">exit</span>
--evalue EVALUE                     diamond evalue threshold
--maxtargetseqs MAXTARGETSEQS       max target sequences to retrieve per query
--cpus CPUS                         number of cpus to use
--tmpdir TMPDIR                     &lt;/path/to/tmp/directory&gt;
--top-pct TOP_PCT                   top percentage of hits to retrieve
--force                             force overwrite of diamond output table
--verbose                           add verbosity
</pre></div>
</div>
<p class="rubric">Functions</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#diamond.DiamondResult" title="diamond.DiamondResult"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diamond.DiamondResult</span></code></a>(qseqid, sseqid, …)</p></td>
<td><p>docstring for DiamondResult.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-diamond"></span><p>Class and functions related to running diamond on metagenome sequences</p>
<dl class="class">
<dt id="diamond.DiamondResult">
<em class="property">class </em><code class="sig-prename descclassname">diamond.</code><code class="sig-name descname">DiamondResult</code><span class="sig-paren">(</span><em class="sig-param">qseqid</em>, <em class="sig-param">sseqid</em>, <em class="sig-param">pident</em>, <em class="sig-param">length</em>, <em class="sig-param">mismatch</em>, <em class="sig-param">gapopen</em>, <em class="sig-param">qstart</em>, <em class="sig-param">qend</em>, <em class="sig-param">sstart</em>, <em class="sig-param">send</em>, <em class="sig-param">evalue</em>, <em class="sig-param">bitscore</em><span class="sig-paren">)</span><a class="headerlink" href="#diamond.DiamondResult" title="Permalink to this definition">¶</a></dt>
<dd><p>docstring for DiamondResult.</p>
<p>Some operator overloading here…
For other examples of this see:</p>
<blockquote>
<div><p><a class="reference external" href="https://www.geeksforgeeks.org/operator-overloading-in-python/">https://www.geeksforgeeks.org/operator-overloading-in-python/</a></p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="diamond.add_taxids">
<code class="sig-prename descclassname">diamond.</code><code class="sig-name descname">add_taxids</code><span class="sig-paren">(</span><em class="sig-param">hits</em>, <em class="sig-param">database</em>, <em class="sig-param">verbose=True</em><span class="sig-paren">)</span><a class="headerlink" href="#diamond.add_taxids" title="Permalink to this definition">¶</a></dt>
<dd><p>Translates accessions to taxid translations from prot.accession2taxid.gz</p>
<p># TODO: Should maybe write a wrapper for this to run on all of the NCBI
databases listed below… Maybe this will help account for instances where the
accession is not found due to suppression,deprecation,removal,etc…</p>
<p>DiamondResults contain sseqids with non-redundant protein sequence database
entries from:
- GenPept
- Swissprot
- PIR
- PDF
- PDB
- RefSeq</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hits</strong> (<em>dict</em>) – {qseqid: DiamondResult, …}</p></li>
<li><p><strong>database</strong> (<em>str</em>) – &lt;/path/to/prot.accession2taxid.gz&gt;</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>hits - {qseqid: DiamondResult, …} sseqids dicts updated with ‘taxid’ key</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – prot.accession2taxid.gz database is required for translation taxid</p></li>
<li><p><strong>DatabasesOutOfDateError</strong> – prot.accession2taxid.gz database and nr.dmnd are out of sync resulting
    in accessions that are no longer available (either due to being
    suppressed, deprecated or removed by NCBI). This must be resolved by
    updating both nr and prot.accession2taxid.gz and re-running diamond on
    the new nr.dmnd database. Alternatively, can try to find the exceptions in merged.dmp</p></li>
</ul>
</dd>
</dl>
<p># TODO: Replace file_length func for database file.
(in this case 808,717,857 lines takes ~15 minutes simply to read each line…)
This causes unnecessary slowdown. On download of the respective database. Line
counts can be saved in config and replaced here for the file_length func. This
will allow for only one read-through of the file, instead of two.</p>
</dd></dl>

<dl class="function">
<dt id="diamond.blast">
<code class="sig-prename descclassname">diamond.</code><code class="sig-name descname">blast</code><span class="sig-paren">(</span><em class="sig-param">fasta</em>, <em class="sig-param">database</em>, <em class="sig-param">outfpath</em>, <em class="sig-param">blast_type='blastp'</em>, <em class="sig-param">evalue=1e-05</em>, <em class="sig-param">maxtargetseqs=200</em>, <em class="sig-param">cpus=0</em>, <em class="sig-param">tmpdir='.'</em>, <em class="sig-param">force=False</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#diamond.blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs diamond blastp search using fasta against diamond formatted database</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta</strong> (<em>str</em>) – &lt;/path/to/fasta/file&gt;. May be amino acid or nucleotide sequences</p></li>
<li><p><strong>database</strong> (<em>str</em>) – &lt;/path/to/diamond/formatted/database&gt;.</p></li>
<li><p><strong>outfpath</strong> (<em>str</em>) – &lt;/path/to/output/file&gt;.</p></li>
<li><p><strong>blast_type</strong> (<em>str</em>) – blastp if fasta consists of amino acids or blastx if nucleotides</p></li>
<li><p><strong>evalue</strong> (<em>float</em>) – cutoff e-value to count hit as significant (the default is float(‘1e-5’)).</p></li>
<li><p><strong>maxtargetseqs</strong> (<em>int</em>) – max number of target sequences to retrieve per query by diamond (the default is 200).</p></li>
<li><p><strong>cpus</strong> (<em>int</em>) – number of cpus to use (the default is 1).</p></li>
<li><p><strong>tmpdir</strong> (<em>type</em>) – &lt;/path/to/temporary/directory&gt; (the default is os.curdir).</p></li>
<li><p><strong>force</strong> (<em>boolean</em>) – overwrite existing diamond results <cite>force</cite> (the default is False).</p></li>
<li><p><strong>verbose</strong> (<em>boolean</em>) – log progress to terminal <cite>verbose</cite> (the default is False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>outfpath</cite></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – <cite>fasta</cite> file does not exist</p></li>
<li><p><strong>ValueError</strong> – provided <cite>blast_type</cite> is not ‘blastp’ or ‘blastx’</p></li>
<li><p><strong>OSError</strong> – Diamond execution failed</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="diamond.parse">
<code class="sig-prename descclassname">diamond.</code><code class="sig-name descname">parse</code><span class="sig-paren">(</span><em class="sig-param">results</em>, <em class="sig-param">top_pct=0.9</em>, <em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#diamond.parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve diamond results from output table</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>results</strong> (<em>str</em>) – &lt;/path/to/blastp/outfmt6/output/file&gt;</p></li>
<li><p><strong>top_pct</strong> (<em>0 &lt; float &lt;= 1</em>) – bitscore filter applied to each qseqid (the default is 0.9).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>{qseqid: DiamondResult, …}</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – diamond results table does not exist</p></li>
<li><p><strong>ValueError</strong> – top_pct value is not a float or not in range of 0 to 1</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hmmer.html" class="btn btn-neutral float-right" title="5.3.1.4. Hmmer" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="bowtie.html" class="btn btn-neutral float-left" title="5.3.1.2. Bowtie" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2015 - 2020, Evan R. Rees, Shaurya Chanana, Siddharth Uppal, Kyle Wolf, Jason C. Kwan
      <span class="lastupdated">
        Last updated on Mar 26, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>